pkg/archive/archive.go

```diff
@@ -148,12 +149,18 @@ func DecompressStream(archive io.Reader) (io.ReadCloser, error) {
 		readBufWrapper := p.NewReadCloserWrapper(buf, bz2Reader)
 		return readBufWrapper, nil
 	case Xz:
-		xzReader, err := xzDecompress(buf)
+		xzReader, chdone, err := xzDecompress(buf)
 		if err != nil {
 			return nil, err
 		}
 		readBufWrapper := p.NewReadCloserWrapper(buf, xzReader)
-		return readBufWrapper, nil
+		return ioutils.NewReadCloserWrapper(readBufWrapper, func() error {
+			go func() {
```

Comment:
I think above comments already has the root cause clear already:

1. docker-untar write pipe will be closed if docker-untar fail, it is using a bufio from pools, and pools.Put(bufio) be called in p.ReadCloserWrapper too early, while the goroutine in exec.Cmd is still doing io.Copy

because chrootarchive/archive.untarHandler has a `defer decompressedArchive.Close()` that is to be called when untarHandler returns, that Close will call pools.Put(buf) which calls buf.Reset(nil) resets that bufio's internal pointers; while exec.Cmd still has the goroutine doing io.Copy, that is going to crash on L86 of bufio.fill
https://github.com/docker/docker/blob/master/pkg/chrootarchive/archive.go#L52-L59
https://golang.org/src/bufio/bufio.go#L83-89

the solution is to wait cmd.Wait finish, that's the point exec.Cmd internal goroutines has all finished.